MFA using an Authenticator app

dantownsend opened the pull request:
If you look at `example_projects/mfa_demo` there's an example project, where MFA using an authenticator app works.

The authenticator MFA backend is pretty much done.

The email MFA is only half done - I might remove it from this PR.

Remaining tasks:

- [x] Finish tests
- [x] Fix linter issues
- [x] Update docs
- [x] Remove email provider for now?

sinisaos commented:
@dantownsend I just tried the `mfa_demo` with Authy. Piccolo app is successfully registered with Authy and code generated successfully and everything works, but I'm not clear on something. Sorry if I don't understand well, but I can login without TOTP code because `session_login()` created a session cookie. Shouldn't this happen if the `mfa_providers` argument is specified in `session_login(mfa_providers=[AuthenticatorProvider()])`. I skipped the `sign up for MFA` step and was still able to access the protected page like in the video?

[mfa.webm](https://github.com/user-attachments/assets/1e6f8711-ce32-41a1-a60d-0f9a93f9881e)


Shouldn't we somehow force the user to first register the TOTP app and then go to re-login with the TOTP code if we have `mfa_providers` argument in `session_login()`?

dantownsend commented:
@sinisaos Thanks for trying it.

At the moment, it doesn't force the user to setup MFA. So if they haven't yet, it just lets them login as usual with username / password.

We could add an option which forces them to set it up.

Also, if someone is logged in, and then sets up MFA, it doesn't invalidate their previous session cookie at the moment, meaning they can still access private pages. We might want to invalidate their previous session cookie ... I'm not sure if OWASP has recommendations on this.

sinisaos commented:
@dantownsend Thanks for the explanation. Looking ahead, this actually makes sense because when this is implemented in Piccolo Admin, the admin user **first login** attempt will be with `username/password` only, and then if I want `MFA`, they will have to enable it (register MFA) in the admin UI. After that, they will have to log in with the TOTP code (login form with TOTP field) every time they try to log in again. 
If the user does not want `MFA`, it will simply not enable it (or disable it if previously enabled) and the user will be able to log in only with `username/password` as we did before. 
Did I understand correctly?

dantownsend commented:
@sinisaos Yes, that's right.

Skelmis commented:
> Also, if someone is logged in, and then sets up MFA, it doesn't invalidate their previous session cookie at the moment, meaning they can still access private pages. We might want to invalidate their previous session cookie ... I'm not sure if OWASP has recommendations on this.

Yea, so in terms of this I'd definitely advise going down the route of invalidating previous session cookies when MFA get's configured. If you didn't want to impact the 'current real user', you could always have it invalidate all sessions bar the one that was used to configure the MFA as that's "trusted" since it's run the privileged access.

---

I also like the idea of supporting MFA enforcement through the usage of an option. Possibly even it could be configured such that maybe you only enforce MFA for admins? Just thinking out loud for end users, there may be a case that it's worthwhile ensuring admins have MFA set up to view items such as the admin portal, but it's too cumbersome for a regular user flow on the application. So possible flags such as `enforce_mfa_for_admins` and `enforce_mfa_for_everyone` come to mind

---

I would also like to just note that once Piccolo goes down the route of supporting MFA as a built in feature, there will also need to be a built in way to reset said MFA in the case access is lost. Often this comes in the form of single-use recovery codes which can be used to reset the MFA if required. Obligatory [OWASP link](https://cheatsheetseries.owasp.org/cheatsheets/Multifactor_Authentication_Cheat_Sheet.html#resetting-mfa)

dantownsend commented:
@Skelmis Thanks for that - it's very useful. 👍

sinisaos commented:
@dantownsend Sorry to bother again, but it is based on the excellent link on OWASP given in previous comments. The user should use a backup (Authy has a [backup in case the phone is lost)](https://authy.com/features/backup/), but we don't know that the user will do that. 
Should we provide some sort of recovery code to make sure the user can access account if they lose their phone etc? 
If so, maybe we could add a `recovery_code` column in the `AuthenticatorSecret` table and then use that column in authenticate method. I don't know if it's good idea and safe enough, but it could be something like this
<details>
<summary>Example code</summary>

```python
# mfa/authenticator/tables.py
class AuthenticatorSecret(Table):
  ...
  recovery_code = Text(secret=True)
  ...

  @classmethod
  def generate_recovery_code(cls) -> str:
      # should we hash it like passwords?
      return secrets.token_hex(16)

  @classmethod
  async def create_new(cls, user_id: int) -> AuthenticatorSecret:
      instance = cls(
          {
              ...
              cls.recovery_code: cls.generate_recovery_code(),
          }
      )
  @classmethod
  async def authenticate(cls, user_id: int, code: str) -> bool:
      ...
      for secret in secrets:
          ...
          if totp.verify(code) or secret.recovery_code == code:
          ...  
```
and then in provider.py
```python
async def get_registration_html(self, user: BaseUser) -> str:
    """
    When a user wants to register for MFA, this HTML is shown containing
    instructions.
    """
    qrcode_image = await self._generate_qrcode_image(user=user)  # noqa
    recovery_code = (
        await self.secret_table.objects()
        .where(user.id == self.secret_table.user_id)
        .first()
    )

    return f"""
        <p>Use an authenticator app like Google Authenticator to scan this QR code:</p>
        <img src="data:image/png;base64,{qrcode_image}" />
        <p>Copy this recovery code and keep it safe:</p>
        <input type="text" value={recovery_code.recovery_code}>
        """  # noqa: E501

async def get_registration_json(self, user: BaseUser) -> dict:
    """
    When a user wants to register for MFA, the client can request a JSON
    response, rather than HTML, if they want to render the UI themselves.
    """

    qrcode_image = await self._generate_qrcode_image(user=user)
    recovery_code = (
        await self.secret_table.objects()
        .where(user.id == self.secret_table.user_id)
        .first()
    )

    return {
        "qrcode_image": qrcode_image,
        "recovery_code": recovery_code.recovery_code,
    }
```

</details>

Skelmis commented:
Oooo nice nice! I'd definitely want to treat the recovery code like a password (it's essentially a more powerful version of one after all), but I don't know what a 'nice' user flow would look like?

It feels like either:
- `create_new` needs to return `tuple[..., str]` where the `str` is the cleartext recovery code *or* the model needs some form of attribute set which shows the recovery code unhashed if it's a new object and null otherwise
- Or there needs to be a second method such as `create_new_code` which stores the hashed code in the model and returns the unhashed code so it can be displayed. My issue with this though is it seems like it'll introduce an extra SQL call which while negligible, is still an extra call every time by design

sinisaos commented:
@Skelmis Thanks. It's really great to have a security expert who can give the best advice. I've only posted sample code and I'm sure Daniel would provide a much better implementation than me. Thanks again for the valuable advice.

sinisaos commented:
To hash the recovery code, we can reuse the Piccolo password code. Something like this. Hope is useful
<details>
<summary>Example code</summary>

```python
# in mfa/utils.py
import hashlib
import secrets

def hash_recovery_code(
    code: str, salt: str = secrets.token_hex(16), iterations: int = 600_000
) -> str:
    hashed = hashlib.pbkdf2_hmac(
        "sha256",
        bytes(code, encoding="utf-8"),
        bytes(salt, encoding="utf-8"),
        iterations,
    ).hex()
    return f"{iterations}${salt}${hashed}"


def split_stored_recovery_code(code: str) -> t.List[str]:
    elements = code.split("$")
    if len(elements) != 3:
        raise ValueError("Unable to split recovery_code")
    return elements


def verify_recovery_code(code: str, stored_code: str) -> bool:
    iterations, salt, _ = split_stored_recovery_code(stored_code)
    return hash_recovery_code(code, salt, int(iterations)) == stored_code

# in mfa/authenticator/tables.py
from __future__ import annotations

import datetime
import logging
import secrets
import typing as t

from piccolo.columns import Integer, Serial, Text, Timestamptz, Varchar
from piccolo.table import Table
from .utils import hash_recovery_code, verify_recovery_code
...
RECOVERY_CODE = secrets.token_hex(16)
...
class AuthenticatorSecret(Table):
    ...
    recovery_code = Text(secret=True)

    @classmethod
    def generate_recovery_code(cls) -> str:
        return RECOVERY_CODE

    @classmethod
    async def create_new(cls, user_id: int) -> AuthenticatorSecret:
        instance = cls(
            {
                cls.user_id: user_id,
                cls.secret: cls.generate_secret(),
                cls.recovery_code: hash_recovery_code(RECOVERY_CODE),
            }
        )
        await instance.save()
        return instance

    @classmethod
    async def authenticate(cls, user_id: int, code: str) -> bool:
        ...
            # check TOTP code or recovery code
            if totp.verify(code) or verify_recovery_code(
                code, secret.recovery_code
            ):
           ...     

# in mfa/provider.py
class AuthenticatorProvider(MFAProvider):
    ...
    async def get_registration_html(self, user: BaseUser) -> str:
        """
        When a user wants to register for MFA, this HTML is shown containing
        instructions.
        """
        qrcode_image = await self._generate_qrcode_image(user=user)  # noqa
        recovery_code = self.secret_table.generate_recovery_code()

        return f"""
            <p>Use an authenticator app like Google Authenticator to scan this QR code:</p>
            <img src="data:image/png;base64,{qrcode_image}" />
            <p>Copy this recovery code and keep it safe:</p>
            <input type="text" value={recovery_code}>
            """  # noqa: E501

    async def get_registration_json(self, user: BaseUser) -> dict:
        """
        When a user wants to register for MFA, the client can request a JSON
        response, rather than HTML, if they want to render the UI themselves.
        """

        qrcode_image = await self._generate_qrcode_image(user=user)
        recovery_code = self.secret_table.generate_recovery_code()

        return {
            "qrcode_image": qrcode_image,
            "recovery_code": recovery_code,
        }
```

</details>

dantownsend commented:
Thanks both - this is a very useful discussion.

@sinisaos Your code looks great - thanks. I couldn't find any particular spec for recovery codes, so I had a look at what other apps are doing.

- Some seem to use UUID v4, which I think is slightly overkill
- Sentry and GitHub do something like `ABC1-XYZ9`. I quite like the `-` separator making things easier to read.
- Almost everyone uses alphanumeric characters, and presumably the recovery codes are case insensitive.

I see in your code you use `secrets.token_hex(16)` which I think is a nice solution. The only issue with hex encoding is it's not the most compact, so we could do something like this:

```python
>>> import secrets
>>> import string
>>> "".join(secrets.choice(string.ascii_lowercase + string.digits) for _ in range(12))
```

I looked at work flows for using recovery codes, and the simplest is what GitLab does - the form lets you enter your TOTP code, or a recovery code:

<img width="495" alt="Screenshot 2024-08-13 at 05 27 43" src="https://github.com/user-attachments/assets/3c5fbbdc-1ae2-41d5-addc-413748afc919">

It looks like this is how you solved it in your code too. We just need to make sure we revoke the recovery code once it has been used.

Most of the apps I checked give multiple recovery codes (somewhere between 8 and 16). I think 8 is more than enough for something like Piccolo Admin. It does mean we'll need a separate table for storing recovery codes. An array column kind of works, but we need to track when a recovery code was used.

Skelmis commented:
In terms of tracking if a code is used, could we just remove it from the array?

And then if/once a user gets down to only a couple codes left we can prompt them to create a new set of codes.

dantownsend commented:
@Skelmis Yes, that's a good idea. If we wanted to track when a user used recovery codes, we could always append a timestamp to another array.

Skelmis commented:
Ooo yes. Do array types support tuples or a more complex data structure? Just thinking about how useful a 'has been used' value is on its own compared to alongside some other metadata such as used_at

sinisaos commented:
I have one question. If we store the recovery codes as plain text in the database, then we can use `8-16` recovery codes, but if we store `8-16 hashed` recovery codes, that would be a problem because validating those hashes will take too long on each login (we will have to check those codes on every login attempt because the recovery code is just as valid as the `TOTP` code)?

dantownsend commented:
> I have one question. If we store the recovery codes as plain text in the database, then we can use 8-16 recovery codes, but if we store 8-16 hashed recovery codes, that would be a problem because validating those hashes will take too long on each login (we will have to check those codes on every login attempt because the recovery code is just as valid as the TOTP code)?

@sinisaos I was thinking the same. If we make the salt and iterations the same for each of the user's recovery keys we can hash the security code provided by the user, and compare it to all of the hashed recovery keys stored in the database.

dantownsend commented:
> Ooo yes. Do array types support tuples or a more complex data structure? Just thinking about how useful a 'has been used' value is on its own compared to alongside some other metadata such as used_at

@Skelmis We can do multidimensional arrays, or use a JSON column. I think we need to store somewhere that they logged in using a recovery code, just not sure where yet.

Skelmis commented:
I have a couple ideas, so I may create some psudo code after dinner. What is Piccolos appetite for creating new tables to ship with? @dantownsend

dantownsend commented:
@Skelmis I'm open to adding new tables. Were you thinking of something like a login history table or something?

Skelmis commented:
Yea, something along those lines but hopefully more generalized such that it could be considered a start for the [audit log](https://github.com/piccolo-orm/piccolo_admin/issues/51) issue which is open. Possibly something along the lines of:
```py
from enum import Enum, auto

from piccolo.apps.user.tables import BaseUser
from piccolo.columns import (
    Serial,
    Text,
    Timestamptz,
    ForeignKey,
    OnDelete,
    OnUpdate,
    JSON,
)
from piccolo.table import Table


class AuditActions(Enum):
    LOGIN = auto()
    RECOVERY_CODE_USED = auto()


class Auditing(Table):
    id: Serial
    action = Text(
        index=True,
        choices=AuditActions,
        help_text="The type of action which occurred.",
    )
    occurred_at = Timestamptz(help_text="When this action occurred.")
    done_by = ForeignKey(
        BaseUser,
        index=True,
        on_delete=OnDelete.no_action,
        on_update=OnUpdate.no_action,
        help_text="Where applicable, the user who triggered this event.",
    )
    description = Text(help_text="A more detailed description of the triggered event.")
    # JSON to allow flexibility within future events
    metadata = JSON(help_text="Metadata specific to this event")
```

Such that it also remains easily usable for various types of auditing actions

sinisaos commented:
> @sinisaos I was thinking the same. If we make the salt and iterations the same for each of the user's recovery codes, when the user tries logging in with a recovery code we just need to hash it once, and compare it to all of the hashed recovery codes stored in the database.

@dantownsend I slightly modified my code from the previous comment (hashing the recovery code) like this to hash and verify `8 recovery codes` on login attempt
<details>
<summary>Example code</summary>

```python
# in mfa/authenticator/tables.py
from .recovery_codes import generate_recovery_code

RECOVERY_CODE = [generate_recovery_code() for _ in range(8)]


class AuthenticatorSecret(Table):
    ...
    recovery_code = Text(secret=True)

    @classmethod
    def generate_recovery_code(cls) -> str:
        return "\n".join([i for i in RECOVERY_CODE])

    @classmethod
    async def create_new(cls, user_id: int) -> AuthenticatorSecret:
        instance = cls(
            {
                cls.user_id: user_id,
                cls.secret: cls.generate_secret(),
                cls.recovery_code: ",".join(
                    [hash_recovery_code(i) for i in RECOVERY_CODE]
                ),
            }
        )
        await instance.save()
        return instance

    @classmethod
    async def authenticate(cls, user_id: int, code: str) -> bool:
        ... # old code
        secret_recovery_codes = secret.recovery_code.split(",")
        verified = [
            i
            for i in secret_recovery_codes
            if verify_recovery_code(code, i) is True
        ]

        if totp.verify(code) or len(verified) > 0:
            secret.last_used_at = datetime.datetime.now(
                tz=datetime.timezone.utc
            )
         ... # old code

# in mfa/authenticator/provider.py
async def get_registration_html(self, user: BaseUser) -> str:
    """
    When a user wants to register for MFA, this HTML is shown containing
    instructions.
    """
    qrcode_image = await self._generate_qrcode_image(user=user)  # noqa
    recovery_code = self.secret_table.generate_recovery_code()

    return f"""
        <p>Use an authenticator app like Google Authenticator to scan this QR code:</p>
        <img src="data:image/png;base64,{qrcode_image}" />
        <p>Copy this recovery code and keep it safe:</p>
        <textarea rows="8" cols="50">{recovery_code}</textarea>
        """  # noqa: E501

async def get_registration_json(self, user: BaseUser) -> dict:
    """
    When a user wants to register for MFA, the client can request a JSON
    response, rather than HTML, if they want to render the UI themselves.
    """

    qrcode_image = await self._generate_qrcode_image(user=user)
    recovery_code = self.secret_table.generate_recovery_code()

    return {
        "qrcode_image": qrcode_image,
        "recovery_code": recovery_code,
    }
```
</details>

But it takes more than `6 seconds` on localhost. I think it's taking too long for every login attempt.

![response](https://github.com/user-attachments/assets/6a1e485d-0a23-4ae3-b345-ae3a1f6b810c)

Skelmis commented:
I think this is because the current method attempts to rehash `code` before it compares it to each `stored_code` where in practice you could get away with only hashing it once possibly try something like this?
```py
    @classmethod
    async def authenticate(cls, user_id: int, code: str) -> bool:
        ... # old code
        secret_recovery_codes = secret.recovery_code.split(",")
        iterations, salt, _ = split_stored_recovery_code(code)
        hashed_code = hash_recovery_code(code, salt, int(iterations))

        if totp.verify(code) or hashed_code in secret_recovery_codes:
            secret.last_used_at = datetime.datetime.now(
                tz=datetime.timezone.utc
            )
         ... # old code
```

dantownsend commented:
@sinisaos Yeah, I've pushed some changes now which does what @Skelmis has outlined above - it just hashes the recovery code provided by the user once, and then compares it to all of the stored recovery codes in the database.

I've updated the migrations, so you would need to rerun them, but if you have time see if it's any faster.

sinisaos commented:
> I think this is because the current method attempts to rehash `code` before it compares it to each `stored_code` where in practice you could get away with only hashing it once, possibly try something like this and see if it helps?
> 
> ```python
>     @classmethod
>     async def authenticate(cls, user_id: int, code: str) -> bool:
>         ... # old code
>         secret_recovery_codes = secret.recovery_code.split(",")
>         iterations, salt, _ = split_stored_recovery_code(code) # we need to check the stored hash code from the database, not form input
>         hashed_code = hash_recovery_code(code, salt, int(iterations))
> 
>         if totp.verify(code) or hashed_code in secret_recovery_codes:
>             secret.last_used_at = datetime.datetime.now(
>                 tz=datetime.timezone.utc
>             )
>          ... # old code
> ```

@Skelmis Thanks for the tip, but it doesn't work because we need to check the stored hash code from the database, not the code from the form input. To make it work, I have to make these changes. Sorry if I missed your point.
```python
results = []
for i in secret.recovery_code.split(","):
    iterations, salt, _ = split_stored_recovery_code(i)
    hashed_code = hash_recovery_code(code, salt, int(iterations))
    results.append(hashed_code)

if totp.verify(code) or hashed_code in results:
```
and the result is the same (response time is over `6 seconds`). But that doesn't matter because Daniel made some changes so I'll try them now to see if it makes a difference.

sinisaos commented:
@dantownsend I tried your changes and they are great. The response is now `2 seconds` which is much better.

sinisaos commented on piccolo_api/mfa/authenticator/provider.py:
```diff
@@ -0,0 +1,93 @@
+        qrcode_image = await self._generate_qrcode_image(
+            secret=secret, email=user.email
+        )
+
```
I think you should add it `recovery_codes_str = "\n".join(recovery_codes)` so we get a nice recovery code format on the json response as well.

![admin_recovery_codes](https://github.com/user-attachments/assets/d0507f91-0b22-4a67-9a0c-b13a84c4d281)

I've updated the [my Piccolo Admin branch](https://github.com/piccolo-orm/piccolo_admin/compare/master...sinisaos:piccolo_admin:two_factor_auth) to use the code from this PR. If you find something interesting and the code is good enough, feel free to use it.

dantownsend replied:
> I think you should add it recovery_codes_str = "\n".join(recovery_codes) so we get a nice recovery code format on the json response as well.

Yeah, it's a tricky one - in some ways having it pre-formatted is convenient, but in others the client might want more control over how it appears (e.g. putting it in `<ul>` tag).

> I've updated the [my Piccolo Admin branch](https://github.com/piccolo-orm/piccolo_admin/compare/master...sinisaos:piccolo_admin:two_factor_auth) to use the code from this PR. If you find something interesting and the code is good enough, feel free to use it.

That's great - thanks! I'll have a look.

dantownsend commented:
I think this is almost done now.

@Skelmis What's your opinion on encrypting the shared secret in the database? Obviously we can't hash it, because we need access to the original shared secret.

At the moment I'm using the Python `cryptography` library - using the Fernet algorithm to encrypt it before storing it in the database. The encryption / decryption key is just a random string provided by the app developer. To gain access, a hacker would need access to the database and the code to get the encryption key.

Do you think it's worth doing?

Skelmis commented:
> I think this is almost done now.
> 
> @Skelmis What's your opinion on encrypting the shared secret in the database? Obviously we can't hash it, because we need access to the original shared secret.
> 
> At the moment I'm using the Python `cryptography` library - using the Fernet algorithm to encrypt it before storing it in the database. The encryption / decryption key is just a random string provided by the app developer. To gain access, a hacker would need access to the database and the code to get the encryption key.
> 
> Do you think it's worth doing?

Definitely worth encrypting that secret at rest, and ensuring that the key used to encrypt that value comes from something such as an environment variable such that it isn't stored alongside the encrypted value. While somewhat lacking I think the OWASP section that applies here can be found [here](https://cheatsheetseries.owasp.org/cheatsheets/Secrets_Management_Cheat_Sheet.html#7-encryption).

The takeaway being that `Fernet` appears to be `AES-128` in `CBC` mode ([Source](https://cryptography.io/en/latest/fernet/#implementation)) which seems to be beneath the recommendation for `AES-256` in `GCM` mode. Given we have the ability to decide before anything is released (at the cost of re-implementing a couple methods), it's likely a good idea to move to something else which is considered recommended. As per OWASP, this would be "`AES-256 using GCM, or a mixture of ChaCha20 and Poly1305`". 

As it currently stands, this will be by using [PyNaCl](https://pynacl.readthedocs.io/en/latest/) which binds libsodium. Looking into this further it seems we may wish to go down the route of `XChaCha20-Poly1305` ([1](https://github.com/pyca/pynacl/issues/515),[2](https://doc.libsodium.org/secret-key_cryptography/aead/aes-256-gcm#warning),[3](https://soatok.blog/2020/05/13/why-aes-gcm-sucks/)), further the AES route does appear within the existing `cryptography` library [4](https://cryptography.io/en/latest/hazmat/primitives/aead/#cryptography.hazmat.primitives.ciphers.aead.AESGCM) but it is only exposed as a primitive which isn't the most ideal use case. 

As such my proposal for this would be using `XChaCha20` and `Poly1305`. This can be implemented via [Aead within pynacl](https://pynacl.readthedocs.io/en/latest/secret/#example-with-aead) which does encryption via a `XChaCha20 stream cipher` and authentication using `Poly1305`. As a part of this, `cryptography` can be removed as a dependency assuming its not required by something else. An example implementation of this may look like the following:
```python
def get_cryptography() -> nacl:
    try:
        import nacl
    except ImportError as e:
        print("Install pip install piccolo_api[authenticator] to use this feature.")
        raise e

    return nacl


def _generate_secure_key() -> bytes:
    nacl = get_cryptography()
    return nacl.utils.random(nacl.secret.Aead.KEY_SIZE)


def _get_nacl_box(secret: bytes) -> nacl.secret.Aead:
    nacl = get_cryptography()
    return nacl.secret.Aead(secret)


def _encrypt(value: str, db_encryption_key: str) -> str:
    box = _get_nacl_box(db_encryption_key)
    encrypted_value = box.encrypt(value)
    return encrypted_value.decode("utf8")


def _decrypt(encrypted_value: str, db_encryption_key: str) -> str:
    box = _get_nacl_box(db_encryption_key)
    decrypted_value = box.decrypt(encrypted_value)
    return decrypted_value.decode("utf8")
```

Although I am having a couple grumbles with pynacl versions which are currently installed on my box versus what the documentation states... I.e. the code above comes from the documentation but doesn't run at face value although I believe to simply be a local problem with my current dev machine as the relevant methods *are* installed and viewable, Python just doesn't believe so.

---

Further, I would have a secondary recommendation and that's we (Piccolo) expose a crypto module. I'd propose essentially a `piccolo[cryptography]` install which installs PyNaCl and allows users to securely encrypt and decrypt things utilizing the same flow we use. The reason I raise this for consideration is that if end users wish to implement their own authentication providers they could simply hook into the existing flows which are already defined and installed to support MFA. As such, we could simply include a link in the documentation such that as a flow for creating providers you use say `piccolo.crypto` to store your secret values in the same way Piccolo does internally. In this, we could easily prompt users to actually treat secrets, as well secrets, instead of forcing them to either investigate a good encryption flow or simply store it in plaintext.

I will note this is a recommendation which could go either way and doesn't bother me regardless, as it isn't necessarily Piccolo's place in the chain to push users down a specific path so to speak. I just think it may possibly be nice to simply publicly expose methods that are in place anyway and have had the opportunity to be peer review by the parties involved with this. Regardless, I'm keen to hear your thoughts

---

As side reading if you are interested, Soatok has some nice blog posts deep diving into cryptographic things.
- A Furry’s Guide to End-to-End Encryption - [Link](https://soatok.blog/2020/11/14/going-bark-a-furrys-guide-to-end-to-end-encryption/)
- Why AES-GCM Sucks - [Link](https://soatok.blog/2020/05/13/why-aes-gcm-sucks/) (This is also ref `3`)
- Comparison of Symmetric Encryption Methods - [Link](https://soatok.blog/2020/07/12/comparison-of-symmetric-encryption-methods/)

Skelmis commented on piccolo_api/mfa/authenticator/tables.py:
```diff
@@ -0,0 +1,270 @@
+        )
+        totp = pyotp.TOTP(shared_secret)  # type: ignore
+
+        if totp.verify(code):
```
Consideration should be given to the `valid_window` argument in this method. This is fine as is, but it does slightly reduce usability if end users have a slight time delay between their clock and the server clock. Further, as this method will take a bit of time (miniscule but still) to do things such as decrypt the secret before verification it ever so slightly increases the chance that by the time MFA is verified it has expired. Bonus points if users submit the totp code within the last bit a code is valid on their phone or they are on the other side of the world to the application in question.

I will note that the current implementation is fine, and also means `last_used_code` is fine with a single `Text` entry. If Piccolo were to support a more conventional window (Current + one past + one future) then the `last_used_code` would need to essentially support three codes at once instead of one due to three codes being valid at any given time. Adding support for that does introduce a touch of complexity, but is something to consider from a usability side.

Here is an [RFC link](https://datatracker.ietf.org/doc/html/rfc6238#section-5.2) on validation and window sizes.

Again, feel free to resolve this point. Either of the two approaches is sufficient but I just wanted to raise it as a discussion point so you were aware of the implications.

dantownsend replied:
Yeah, it's tricky.

We could have a param to `MFAProvider` where the user can specify the `valid_window`.

With `last_used_code`, the logic does get quite tricky if we start storing multiple codes. It's to guard against replay attacks - presumably the vulnerability is if someone logs in twice in quick succession, and then if the first code is replayed it is still valid.

I'm not sure - what do you think?

Skelmis replied:
Yea, it is unfortunately one of those areas where I feel the cost to implement it is non linear haha.

Honestly for now it's probably just easier to leave as is, with a single valid token at any given time. If users do have issues they could also raise it and we could resolve it at a later date rather than adding that cost factor to this pr.

Both ways do work and are fairly equally valid choices I feel, in this case I'd probably err on the side of what's already implemented and just have it as a discussion point for possibly some time in the future

dantownsend replied:
OK, I'll add a `valid_window` argument for now, and will mention the pros and cons of increasing it. Thanks!

Skelmis reviewed (changes requested):
There seems to be some implicit differences where about half the code looks to support multiple devices per authenticator (I.e. totp codes on two phones with different secrets) while the other half looks to only allow one secret per user. I've added some comments where I saw it, but if you could clarify your expectations I can go resolve some stuff or add appropriate comments.

Otherwise I've just added a few comments here and there but overall it looks darn good!! Looking forward to seeing it continue coming together

dantownsend commented on piccolo_api/mfa/authenticator/provider.py:
```diff
@@ -0,0 +1,143 @@
+        Allows authentication using an authenticator app on the user's phone,
+        like Google Authenticator.
+
+        :param db_encryption_key:
+            The shared secrets are encrypted in the database - pass in a random
+            string which is used for encrypting them.
```
That's a good point - I checked, and pycryto doesn't allow any arbitrary string, it requires a specific method on pycrypto to be called to generate a suitable key.

dantownsend commented on piccolo_api/mfa/authenticator/tables.py:
```diff
@@ -0,0 +1,270 @@
+        help_text="Whenever a recovery code is used, store a timestamp here.",
+    )
+    created_at = Timestamptz()
+    revoked_at = Timestamptz(null=True, default=None)
```
Good point - thanks!

dantownsend commented on piccolo_api/mfa/authenticator/tables.py:
```diff
@@ -0,0 +1,270 @@
+    async def authenticate(
+        cls, user_id: int, code: str, db_encryption_key: str
+    ) -> bool:
+        secret = (
+            await cls.objects()
+            .where(
+                cls.user_id == user_id,
+                cls.revoked_at.is_null(),
+            )
+            .order_by(cls.created_at, ascending=False)
+            .first()
+        )
+
```
My original intent was to allow the user to register multiple devices, but then it adds complexity, because the user needs a way of identifying the device, if they want to disable it.

dantownsend commented on piccolo_api/mfa/authenticator/tables.py:
```diff
@@ -0,0 +1,270 @@
+class AuthenticatorSecret(Table):
+    id: Serial  # TODO - we might change this to a UUID primary key
+    user_id = Integer(null=False)
+    device_name = Varchar(
+        null=True,
+        default=None,
+        help_text=(
+            "The user can specify this to make the device memorable, "
+            "if we want to allow them to delete secrets."
+        ),
+    )
```
You're right - this isn't being used at the moment. I'll either add it as an field in the UI, or just remove it.

dantownsend commented on piccolo_api/mfa/authenticator/tables.py:
```diff
@@ -0,0 +1,270 @@
+
+        return (instance, recovery_codes)
+
+    @classmethod
+    async def revoke_all(cls, user_id: int):
+        now = datetime.datetime.now(tz=datetime.timezone.utc)
+        await cls.update({cls.revoked_at: now}).where(
+            cls.user_id == user_id,
+            cls.revoked_at.is_null(),
+        )
```
Yes, good idea!

dantownsend commented on piccolo_api/mfa/email/tables.py:
```diff
@@ -0,0 +1,33 @@
+
+    @classmethod
+    async def create_new(cls, email: str) -> EmailCode:
+        # TODO - generate proper code
```
Good idea.

I had a thought about the email implementation. At the moment it has its own database table, but we could potentially use the `AuthenticatorSecret` table.

My original idea with `EmailCode` is a new one would be generated each time the user logs in.

If instead we used the `AuthenticatorSecret` table, we could generate a secret (like with authenticator apps), but rather than sharing that secret with the client, each time they login we just generate a TOTP code using the secret and send it to them via email. When the client submits the code, we verify it using the secret.

dantownsend commented on piccolo_api/mfa/endpoints.py:
```diff
@@ -0,0 +1,159 @@
+                    return HTMLResponse(content=html_content)
+            elif action == "revoke":
+                if password := body.get("password"):
+                    if await piccolo_user.__class__.login(
+                        username=piccolo_user.username, password=password
+                    ):
```
Good catch. It makes sense to use `self._auth_table` in both cases.

dantownsend replied:
Updated

dantownsend commented on piccolo_api/session_auth/endpoints.py:
```diff
@@ -264,6 +270,67 @@ async def post(self, request: Request) -> Response:
+                            # That could always be in the logic of `send_code`
+                            # though.
+
+                            if return_html:
+                                return self._render_template(
+                                    request,
+                                    template_context={
+                                        "error": "MFA code required",
+                                        "show_mfa_input": True,
+                                        "mfa_token_name": (
+                                            mfa_provider.token_name
+                                        ),
+                                    },
+                                )
+                            else:
+                                raise HTTPException(
+                                    status_code=401, detail="MFA code required"
+                                )
```
Good point!

dantownsend replied:
I've modified the message now, so if any of the providers sent a code, we tell the user about it.

dantownsend commented:
> > I think this is almost done now.
> > @Skelmis What's your opinion on encrypting the shared secret in the database? Obviously we can't hash it, because we need access to the original shared secret.
> > At the moment I'm using the Python `cryptography` library - using the Fernet algorithm to encrypt it before storing it in the database. The encryption / decryption key is just a random string provided by the app developer. To gain access, a hacker would need access to the database and the code to get the encryption key.
> > Do you think it's worth doing?
> 
> Definitely worth encrypting that secret at rest, and ensuring that the key used to encrypt that value comes from something such as an environment variable such that it isn't stored alongside the encrypted value. While somewhat lacking I think the OWASP section that applies here can be found [here](https://cheatsheetseries.owasp.org/cheatsheets/Secrets_Management_Cheat_Sheet.html#7-encryption).
> 
> The takeaway being that `Fernet` appears to be `AES-128` in `CBC` mode ([Source](https://cryptography.io/en/latest/fernet/#implementation)) which seems to be beneath the recommendation for `AES-256` in `GCM` mode. Given we have the ability to decide before anything is released (at the cost of re-implementing a couple methods), it's likely a good idea to move to something else which is considered recommended. As per OWASP, this would be "`AES-256 using GCM, or a mixture of ChaCha20 and Poly1305`".
> 
> As it currently stands, this will be by using [PyNaCl](https://pynacl.readthedocs.io/en/latest/) which binds libsodium. Looking into this further it seems we may wish to go down the route of `XChaCha20-Poly1305` ([1](https://github.com/pyca/pynacl/issues/515),[2](https://doc.libsodium.org/secret-key_cryptography/aead/aes-256-gcm#warning),[3](https://soatok.blog/2020/05/13/why-aes-gcm-sucks/)), further the AES route does appear within the existing `cryptography` library [4](https://cryptography.io/en/latest/hazmat/primitives/aead/#cryptography.hazmat.primitives.ciphers.aead.AESGCM) but it is only exposed as a primitive which isn't the most ideal use case.
> 
> As such my proposal for this would be using `XChaCha20` and `Poly1305`. This can be implemented via [Aead within pynacl](https://pynacl.readthedocs.io/en/latest/secret/#example-with-aead) which does encryption via a `XChaCha20 stream cipher` and authentication using `Poly1305`. As a part of this, `cryptography` can be removed as a dependency assuming its not required by something else. An example implementation of this may look like the following:
> 
> ```python
> def get_cryptography() -> nacl:
>     try:
>         import nacl
>     except ImportError as e:
>         print("Install pip install piccolo_api[authenticator] to use this feature.")
>         raise e
> 
>     return nacl
> 
> 
> def _generate_secure_key() -> bytes:
>     nacl = get_cryptography()
>     return nacl.utils.random(nacl.secret.Aead.KEY_SIZE)
> 
> 
> def _get_nacl_box(secret: bytes) -> nacl.secret.Aead:
>     nacl = get_cryptography()
>     return nacl.secret.Aead(secret)
> 
> 
> def _encrypt(value: str, db_encryption_key: str) -> str:
>     box = _get_nacl_box(db_encryption_key)
>     encrypted_value = box.encrypt(value)
>     return encrypted_value.decode("utf8")
> 
> 
> def _decrypt(encrypted_value: str, db_encryption_key: str) -> str:
>     box = _get_nacl_box(db_encryption_key)
>     decrypted_value = box.decrypt(encrypted_value)
>     return decrypted_value.decode("utf8")
> ```
> 
> Although I am having a couple grumbles with pynacl versions which are currently installed on my box versus what the documentation states... I.e. the code above comes from the documentation but doesn't run at face value although I believe to simply be a local problem with my current dev machine as the relevant methods _are_ installed and viewable, Python just doesn't believe so.
> 
> Further, I would have a secondary recommendation and that's we (Piccolo) expose a crypto module. I'd propose essentially a `piccolo[cryptography]` install which installs PyNaCl and allows users to securely encrypt and decrypt things utilizing the same flow we use. The reason I raise this for consideration is that if end users wish to implement their own authentication providers they could simply hook into the existing flows which are already defined and installed to support MFA. As such, we could simply include a link in the documentation such that as a flow for creating providers you use say `piccolo.crypto` to store your secret values in the same way Piccolo does internally. In this, we could easily prompt users to actually treat secrets, as well secrets, instead of forcing them to either investigate a good encryption flow themselves or simply store it in plaintext.
> 
> I will note this is a recommendation which could go either way and doesn't bother me regardless, as it isn't necessarily Piccolo's place in the chain to push users down a specific path so to speak. I just think it may possibly be nice to simply publicly expose methods that are in place anyway and have had the opportunity to be peer review by the parties involved with this. Regardless, I'm keen to hear your thoughts
> 
> As side reading if you are interested, Soatok has some nice blog posts deep diving into cryptographic things.
> 
> * A Furry’s Guide to End-to-End Encryption - [Link](https://soatok.blog/2020/11/14/going-bark-a-furrys-guide-to-end-to-end-encryption/)
> * Why AES-GCM Sucks - [Link](https://soatok.blog/2020/05/13/why-aes-gcm-sucks/) (This is also ref `3`)
> * Comparison of Symmetric Encryption Methods - [Link](https://soatok.blog/2020/07/12/comparison-of-symmetric-encryption-methods/)

Thanks for this - it's super helpful.

I've refactored some of it today, so there's now an `EncryptionProvider`, so the user can decide implement their own subclass, which uses any encryption library they like.

It's still using Fernet at the moment, but as per your recommendation I'll update it to use `PyNaCI`.

dantownsend commented on piccolo_api/session_auth/endpoints.py:
```diff
@@ -264,6 +270,67 @@ async def post(self, request: Request) -> Response:
+                assert user is not None
+
+                for mfa_provider in mfa_providers:
+                    if await mfa_provider.is_user_enrolled(user=user):
```
That's a good point - thanks.

I went with the second option - if there are multiple MFA providers, a select widget is shown on the login form so the user specifies which they are using.

Skelmis replied:
Sounds good!

dantownsend commented:
OK, I think I've addressed all of the review comments now. Thanks for the really helpful input.

My plan is to release this tomorrow, so I can get the corresponding Piccolo Admin PR passing, and then release that shortly after.

Skelmis commented on piccolo_api/encryption/providers.py:
```diff
@@ -0,0 +1,225 @@
+import typing as t
+from abc import ABCMeta, abstractmethod
+
+import cryptography.fernet
+import nacl.encoding
+import nacl.secret
```
Importing these here negates `get_cryptography` raising a helpful error as line 7 will throw the relevant missing error before a user has the opportunity to hit it

sinisaos replied:
@dantownsend Just one thing. I just tried the Piccolo Admin MFA branch with newest Piccolo API version and removing those imports breaks the MFA flow. The error log is
```bash
File "/home/rkl/dev/admin_env/lib/python3.11/site-packages/piccolo_api/mfa/endpoints.py", line 126, in post
    html_content = await self._provider.get_registration_html(
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "/home/rkl/dev/admin_env/lib/python3.11/site-packages/piccolo_api/mfa/authenticator/provider.py", line 121, in get_registration_html
  secret, recovery_codes = await self.secret_table.create_new(
                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "/home/rkl/dev/admin_env/lib/python3.11/site-packages/piccolo_api/mfa/authenticator/tables.py", line 116, in create_new
  encrypted_secret = encryption_provider.encrypt(value=secret)
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "/home/rkl/dev/admin_env/lib/python3.11/site-packages/piccolo_api/encryption/providers.py", line 122, in encrypt
  fernet = cryptography.fernet.Fernet(  # type: ignore
            ^^^^^^^^^^^^^^^^^^^
AttributeError: module 'cryptography' has no attribute 'fernet'
```
If I patch the local Piccolo API installation with those imports, everything works.

dantownsend replied:
Ah, good catch - I'll do a new release later today.

Skelmis commented:
Just the one final comment otherwise it looks all good to me at face value! Loving the work

dantownsend commented on piccolo_api/mfa/authenticator/utils.py:
```diff
@@ -0,0 +1,33 @@
+    return qrcode
+
+
+def get_b64encoded_qr_image(data):
```
Good catch - I've added type annotations and a docstring.

dantownsend commented:
OK, ready to merge - thanks again!

I think the next improvement I'd love to have is account lockout if a user fails the MFA too many times. But that can be in a future PR.
